src/session_impl.cpp

```diff
@@ -880,13 +878,8 @@ namespace aux {
 		session_log(" aborting all connections (%d)", int(m_connections.size()));
 #endif
 		// abort all connections
-		for (connection_map::iterator i = m_connections.begin();
-			i != m_connections.end();)
-		{
-			peer_connection* p = (*i).get();
-			++i;
+		for (auto& p : m_connections)
 			p->disconnect(errors::stopping_torrent, operation_t::bittorrent);
```

Comment:
this is problematic. the call to `disconnect()` may erase an item from `m_connections`. That's why `i` is incremented before the call.

Reply:
in fact, there should probably be a similar assert for `session_impl` as there is in `torrent` for iterating the peer connections`

Reply:
I thought that with the latest changes, all actual removal of the map are done deferred, am I wrong?

Reply:
oh, yeah, you're right